Implemented Molecule.from_cif(), .from_mol(), .to_mol()

macbiel opened the pull request:
Implemented the following methods:
- Molecule.from_cif(),
- Molecule.from_mol(), and
- Molecule.to_mol().

Implementation notes following.


##### Molecule.from_cif()
- Pretty slow.
- I didn't find a standardized way in which .cif files may include a name for the described molecule. Provisionally, the method  takes the basename of the passed file and uses that as the created molecule's name.
- The .cif file specification allows a fair bit of variety in the way data is presented, but the parser can't exactly cope with it all. As such, there are valid .cif files which the parser will cowardly refuse to deal with and return `None`. Notably:
    1. Some .cif files represent atom coordinates in the `atom_site` loop (table) as absolute coordinates (identified as `cartn_x` etc.), but others use fractional coordinates w.r.t. the unit cell's dimensions (identified as `fract_x` etc.). I don't know exactly how involved conversion between fractional and absolute coordinates would be, but either way, the parser does not carry it out, and instead refuses to parse a file if the `atom_site` loop doesn't have absolute coordinates. [PDB-specific .cif file specifications](https://mmcif.wwpdb.org/docs/tutorials/content/atomic-description.html) appear to mandate using `cartn_x` etc. over `fract_x` etc.
    2. .cif files may represent the element of a given atom in the `atom_site` loop differently. The parser requires the element name to be contained in the `type_symbol` column specifically. PDB-specific .cif file specifications appear to mandate including `type_symbol`.
    3. While the fields in a .cif file loop are whitespace-separated, the specification allows fields to contain whitespace, by putting the entire field in quote - e.g. the record `AUTHOR    'A. Jackson'     1` has 3 fields. Blindly treating each contiguous whitespace as a separator would instead net 4 fields, which is incorrect, and what the parser does. Fortunately, the parser can detect when this has potentially occurred and ceases operation in such cases. For what it's worth, I haven't found any .cif files whose `atom_site` loop actually contained quoted fields.
    4. .cif files provide a way of expressly stating a data value as missing/non-applicable, and this includes an atom's coordinates/element. If the parser encounters a missing value for one of those items, it will refuse to create the molecule.
- Not tested very extensively. The parser can identify and report on the above 4 issues, but if a .cif file has something else that that the parser can't naturally deal with, it'll probably be difficult to debug.


##### Molecule.from_mol()
- Pretty slow.
- Might work for .sdf files too.
- .mol files can carry various information about atoms, including non-standard atomic mass, valence, and implicit hydrogens. The parse ignores most of it - the only atom information other than coordinates and element it makes use of is the atomic charge.
- The 'bond' block can describe bonds of type other than currently representable by the `Bond` class (see page 52 of [this PDF](https://discover.3ds.com/sites/default/files/2020-08/biovia_ctfileformats_2020.pdf). The parser degenerates all of those exotic cases to a single bond.
- The only information extracted from the 'header' block is the first line, which may or may not contain the molecule name.
- The 'properties' block of the .mol file is completely ignored.


##### Molecule.to_mol()
- Given how much information Molecule.from_mol() disregards, parsing a .mol file to a Molecule instance, and then immediately back into a .mol file (i.e. `Molecule.from_mol(file).to_mol()`) is very likely to be lossy.
- Bonds within the 'bond' block might get reordered when compared to the original file, but otherwise remain intact.

Commit:
Implemented Molecule.from_cif(), .from_mol(), .to_mol()

Commit:
Undo mistake

dav3794 commented on psimol/main.py:
```diff
@@ -443,7 +443,57 @@ def from_mol(cls, file_path: str) -> Molecule:
+
+        atom_block = lines[4          : 4+atom_cnt]
+        bond_block = lines[4+atom_cnt : 4+atom_cnt+bond_cnt]
+        print(bond_cnt, len(bond_block))
```
```suggestion
        logging.debug(bond_cnt, len(bond_block))
```
Guess it's rather for debugging.

dav3794 commented on psimol/main.py:
```diff
@@ -452,7 +502,114 @@ def from_cif(cls, file_path: str) -> Molecule:
+            atoms.append(atom)
+
+        # get conjectural name
+        name = file_path.split('/')[-1]
```
```suggestion
        name = os.path.basename(file_path)
```
We can add `import os`, it would be more OS independent.

dav3794 commented on psimol/main.py:
```diff
@@ -443,7 +443,57 @@ def from_mol(cls, file_path: str) -> Molecule:
+                bond_aromaticity = True
+            else:
+                bond_order = 1
+                bond_aromaticity = True
```
Should we treat them as aromatic here?

dav3794 commented on psimol/main.py:
```diff
@@ -491,7 +648,41 @@ def to_mol(self) -> str:
+        for atom in self._atoms:
+            charge = atom.charge if -3 <= atom.charge <= 3 else 0
+            charge = charge_indices[charge]
+            out += f'{atom.x:10.4f}{atom.y:10.4f}{atom.z:10.4f} {atom.symbol:3} 0{charge}  0  0  0  0  0  0  0  0  0\n'
```
```suggestion
            out += f'{atom.x:10.4f}{atom.y:10.4f}{atom.z:10.4f} {atom.symbol:3} 0 {charge}  0  0  0  0  0  0  0  0  0\n'
```
We could remove leading spaces from charge_indices values for better readability

dav3794 commented on psimol/main.py:
```diff
@@ -452,7 +502,114 @@ def from_cif(cls, file_path: str) -> Molecule:
+            line = line.lstrip()
+
+            # phase 1: advance parser to 'atom_site' loop
+            # (in STAR files (CIF included), tables are called "loops")
```
Perhaps we can extract such table with multiline regex from the whole file text (`file.read()`), so possibly we could diminish the number of iterations / make the code more compact, but it's optional.

dav3794 commented on psimol/main.py:
```diff
@@ -452,7 +502,114 @@ def from_cif(cls, file_path: str) -> Molecule:
+                if not in_atom_site_body:
+                    if _is_atom_site_header(line):
+                        # add column name to the list, without the loop name prefix
+                        atom_site_header[line[11:].lower().rstrip()] = ncol
```
Should we keep all the headers other than the ones in `required_colums`?

dav3794 commented on psimol/main.py:
```diff
@@ -452,7 +502,114 @@ def from_cif(cls, file_path: str) -> Molecule:
+            for coord in x, y, z:
+                parenthesis = coord.rfind('(')
+                if parenthesis != -1:
+                    coord = coord[:parenthesis]
```
I don't think it works in-place.

dav3794 reviewed (changes requested):
Added some minor suggestions and possible disambiguations.

dav3794 commented:
If we could, we can also add a method for reading PDB files - their format is very similar to what we already have, so it shouldn't take long to implement (we can support both ATOM and HETATM with CONECT information).

Commit:
Applied suggestions etc.

macbiel commented:
Applied the suggested fixes etc. Comments to each fix below.

1. Yes, that line was a debug message I put there for help with testing. A message of two unannotated numbers is pretty useless, and we don't log debug messages consistently elsewhere, so I just removed the message.
2. Amended as suggested. Also added `import os`.
3. Amended as suggested. Mistake on my part, thank you for noticing.
4. Amended as suggested.
5. Slightly changed how the `atom_site` header is dealt with.
6. Reimplemented the parenthesis removal to a one-liner, using `re`. Also unrolled the loop over `x`, `y`, and `z`.

I'm holding with the implementation of a potential Molecule.from_pdb() method for now.

dav3794 committed:
Merge branch 'main' into main
